File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/IndyException.java
package org.hyperledger.indy.sdk;

import com.sun.jna.ptr.PointerByReference;
import org.hyperledger.indy.sdk.anoncreds.*;
import org.hyperledger.indy.sdk.did.DidAlreadyExistsException;
import org.hyperledger.indy.sdk.ledger.ConsensusException;
import org.hyperledger.indy.sdk.ledger.LedgerInvalidTransactionException;
import org.hyperledger.indy.sdk.ledger.LedgerSecurityException;
import org.hyperledger.indy.sdk.ledger.TimeoutException;
import org.hyperledger.indy.sdk.payments.*;
import org.hyperledger.indy.sdk.pool.*;
import org.hyperledger.indy.sdk.crypto.UnknownCryptoException;
import org.hyperledger.indy.sdk.wallet.*;
import org.json.JSONObject;

/**
 * Thrown when an Indy specific error has occurred.
 */
public class IndyException extends Exception {

	private static final long serialVersionUID = 2650355290834266477L;
	private int sdkErrorCode;
	private String sdkMessage;
	private String sdkBacktrace; // Collecting of backtrace can be enabled by:
								 //   1) setting environment variable `RUST_BACKTRACE=1`
								 //   2) calling `setRuntimeConfig` API function with `collect_backtrace: true`

	/**
	 * Initializes a new IndyException with the specified message.
	 *
	 * @param message The message for the exception.
	 * @param sdkErrorCode The SDK error code to construct the exception from.
	 */
	protected IndyException(String message, int sdkErrorCode) {
		super(message);
		ErrorDetails errorDetails = new ErrorDetails();
		this.sdkErrorCode = sdkErrorCode;
		this.sdkMessage = errorDetails.message;
		this.sdkBacktrace = errorDetails.backtrace;
	}

	/**
	 * Gets the SDK error code for the exception.
	 *
	 * @return The SDK error code used to construct the exception.
	 */
	public int getSdkErrorCode() {
		return sdkErrorCode;
	}

	/**
	 * Gets the SDK error message for the exception.
	 */
	public String getSdkMessage() {
		return sdkMessage;
	}

	/**
	 * Gets the SDK error backtrace for the exception.
	 *
	 * @return The SDK backtrace.
	 */
	public String getSdkBacktrace() {
		return sdkBacktrace;
	}

	private static class ErrorDetails{
		String message;
		String backtrace;

		private ErrorDetails() {
			PointerByReference errorDetailsJson = new PointerByReference();

			LibIndy.api.indy_get_current_error(errorDetailsJson);

			try {
				JSONObject errorDetails = new JSONObject(errorDetailsJson.getValue().getString(0));
				this.message = errorDetails.optString("message");
				this.backtrace = errorDetails.optString("backtrace");
			} catch (Exception ignored){
				// Nothing to do
			}
		}
	}

	/**
	 * Initializes a new IndyException using the specified SDK error code.
	 *
	 * @param sdkErrorCode The SDK error code to construct the exception from.
	 *
	 * @return IndyException correspondent to SDK error code
	 */
	public static IndyException fromSdkError(int sdkErrorCode) {

		ErrorCode errorCode = ErrorCode.valueOf(sdkErrorCode);

		switch (errorCode) {
			case CommonInvalidParam1:
			case CommonInvalidParam2:
			case CommonInvalidParam3:
			case CommonInvalidParam4:
			case CommonInvalidParam5:
			case CommonInvalidParam6:
			case CommonInvalidParam7:
			case CommonInvalidParam8:
			case CommonInvalidParam9:
			case CommonInvalidParam10:
			case CommonInvalidParam11:
			case CommonInvalidParam12:
			case CommonInvalidParam13:
			case CommonInvalidParam14:
				return new InvalidParameterException(sdkErrorCode);
			case CommonInvalidState:
				return new InvalidStateException();
			case CommonInvalidStructure:
				return new InvalidStructureException();
			case CommonIOError:
				return new IOException();
			case WalletInvalidHandle:
				return new InvalidWalletException();
			case WalletUnknownTypeError:
				return new UnknownWalletTypeException();
			case WalletTypeAlreadyRegisteredError:
				return new DuplicateWalletTypeException();
			case WalletAlreadyExistsError:
				return new WalletExistsException();
			case WalletNotFoundError:
				return new WalletNotFoundException();
			case WalletInputError:
				return new WalletInputException();
			case WalletDecodingError:
				return new WalletDecodingException();
			case WalletStorageError:
				return new WalletStorageException();
			case WalletEncryptionError:
				return new WalletEncryptionException();
			case WalletItemNotFound:
				return new WalletItemNotFoundException();
			case WalletItemAlreadyExists:
				return new WalletItemAlreadyExistsException();
			case WalletQueryError:
				return new WalletInvalidQueryException();
			case WalletIncompatiblePoolError:
				return new WrongWalletForPoolException();
			case WalletAlreadyOpenedError:
				return new WalletAlreadyOpenedException();
			case WalletAccessFailed:
				return new WalletAccessFailedException();
			case PoolLedgerNotCreatedError:
				return new PoolConfigNotCreatedException();
			case PoolLedgerInvalidPoolHandle:
				return new InvalidPoolException();
			case PoolLedgerTerminated:
				return new PoolLedgerTerminatedException();
			case LedgerNoConsensusError:
				return new ConsensusException();
			case LedgerInvalidTransaction:
				return new LedgerInvalidTransactionException();
			case LedgerSecurityError:
				return new LedgerSecurityException();
			case PoolLedgerConfigAlreadyExistsError:
				return new PoolLedgerConfigExistsException();
			case PoolLedgerTimeout:
				return new TimeoutException();
			case PoolIncompatibleProtocolVersion:
				return new PoolIncompatibleProtocolVersionException();
			case LedgerNotFound:
				return new LedgerNotFoundException();
			case AnoncredsRevocationRegistryFullError:
				return new RevocationRegistryFullException();
			case AnoncredsInvalidUserRevocId:
				return new AnoncredsInvalidUserRevocId();
			case AnoncredsMasterSecretDuplicateNameError:
				return new DuplicateMasterSecretNameException();
			case AnoncredsProofRejected:
				return new ProofRejectedException();
			case AnoncredsCredentialRevoked:
				return new CredentialRevokedException();
			case AnoncredsCredDefAlreadyExistsError:
				return new CredDefAlreadyExistsException();
			case UnknownCryptoTypeError:
				return new UnknownCryptoException();
			case DidAlreadyExistsError:
				return new DidAlreadyExistsException();
			case UnknownPaymentMethod:
				return new UnknownPaymentMethodException();
			case IncompatiblePaymentError:
				return new IncompatiblePaymentException();
			case InsufficientFundsError:
				return new InsufficientFundsException();
			case ExtraFundsError:
				return new ExtraFundsException();
			case PaymentSourceDoesNotExistError:
				return new PaymentSourceDoesNotExistException();
			case PaymentOperationNotSupportedError:
				return new PaymentOperationNotSupportedException();
			case TransactionNotAllowedError:
				return new TransactionNotAllowedException();
			default:
				String message = String.format("An unmapped error with the code '%s' was returned by the SDK.", sdkErrorCode);
				return new IndyException(message, sdkErrorCode);
		}
	}
}

File: wrappers/dotnet/indy-sdk-dotnet/IndyException.cs
using Hyperledger.Indy.AnonCredsApi;
using Hyperledger.Indy.DidApi;
using Hyperledger.Indy.LedgerApi;
using Hyperledger.Indy.PaymentsApi;
using Hyperledger.Indy.PoolApi;
using Hyperledger.Indy.WalletApi;
using System;

namespace Hyperledger.Indy
{
    /// <summary>
    /// Exception indicating a problem originating from the Indy SDK.
    /// </summary>
    public class IndyException : Exception
    {
        /// <summary>
        /// Initializes a new IndyException with the specified message and SDK error code.
        /// </summary>
        /// <param name="message">The message for the exception.</param>
        /// <param name="sdkErrorCode">The SDK error code for the exception.</param>
        internal IndyException(String message, int sdkErrorCode) : base(message)
        {
            SdkErrorCode = sdkErrorCode;
        }

        /// <summary>
        /// Generates an IndyException or one of its subclasses from the provided SDK error code.
        /// </summary>
        /// <param name="sdkErrorCode">The error code.</param>
        /// <returns>An IndyException or subclass instance.</returns>
        internal static IndyException FromSdkError(int sdkErrorCode)
        {
            var errorCode = (ErrorCode)sdkErrorCode;
            
            switch (errorCode)
            {
                case ErrorCode.CommonInvalidParam1:
                case ErrorCode.CommonInvalidParam2:
                case ErrorCode.CommonInvalidParam3:
                case ErrorCode.CommonInvalidParam4:
                case ErrorCode.CommonInvalidParam5:
                case ErrorCode.CommonInvalidParam6:
                case ErrorCode.CommonInvalidParam7:
                case ErrorCode.CommonInvalidParam8:
                case ErrorCode.CommonInvalidParam9:
                case ErrorCode.CommonInvalidParam10:
                case ErrorCode.CommonInvalidParam11:
                case ErrorCode.CommonInvalidParam12:
                case ErrorCode.CommonInvalidParam13:
                case ErrorCode.CommonInvalidParam14:
                case ErrorCode.CommonInvalidParam15:
                case ErrorCode.CommonInvalidParam16:
                case ErrorCode.CommonInvalidParam17:
                case ErrorCode.CommonInvalidParam18:
                case ErrorCode.CommonInvalidParam19:
                case ErrorCode.CommonInvalidParam20:
                case ErrorCode.CommonInvalidParam21:
                case ErrorCode.CommonInvalidParam22:
                case ErrorCode.CommonInvalidParam23:
                case ErrorCode.CommonInvalidParam24:
                case ErrorCode.CommonInvalidParam25:
                case ErrorCode.CommonInvalidParam26:
                case ErrorCode.CommonInvalidParam27:
                    return new InvalidParameterException(sdkErrorCode);
                case ErrorCode.CommonInvalidState:
                    return new InvalidStateException();
                case ErrorCode.CommonInvalidStructure:
                    return new InvalidStructureException();
                case ErrorCode.CommonIOError:
                    return new IOException();
                case ErrorCode.WalletInvalidHandle:
                    return new InvalidWalletException(); 
                case ErrorCode.WalletUnknownTypeError:
                    return new UnknownWalletTypeException(); 
                case ErrorCode.WalletTypeAlreadyRegisteredError:
                    return new DuplicateWalletTypeException();
                case ErrorCode.WalletAlreadyExistsError:
                    return new WalletExistsException();
                case ErrorCode.WalletNotFoundError:
                    return new WalletNotFoundException();
                case ErrorCode.WalletIncompatiblePoolError:
                    return new WrongWalletForPoolException();
                case ErrorCode.WalletAlreadyOpenedError:
                    return new WalletAlreadyOpenedException();
                case ErrorCode.WalletAccessFailed:
                    return new WalletAccessFailedException();
                case ErrorCode.PoolLedgerNotCreatedError:
                    return new PoolConfigNotCreatedException();
                case ErrorCode.PoolLedgerInvalidPoolHandle:
                    return new InvalidPoolException();
                case ErrorCode.PoolLedgerTerminated:
                    return new PoolLedgerTerminatedException();
                case ErrorCode.PoolIncompatibleProtocolVersionError:
                    return new PoolIncompatibleProtocolVersionException();
                case ErrorCode.PoolLedgerConfigAlreadyExistsError:
                    return new PoolLedgerConfigExistsException();
                case ErrorCode.LedgerNoConsensusError:
                    return new LedgerConsensusException();
                case ErrorCode.LedgerInvalidTransaction:
                    return new InvalidLedgerTransactionException();
                case ErrorCode.LedgerSecurityError:
                    return new LedgerSecurityException();
                case ErrorCode.AnoncredsRevocationRegistryFullError:
                    return new RevocationRegistryFullException();
                case ErrorCode.AnoncredsInvalidUserRevocId:
                    return new InvalidUserRevocIdException();
                case ErrorCode.AnoncredsMasterSecretDuplicateNameError:
                    return new DuplicateMasterSecretNameException();
                case ErrorCode.AnoncredsProofRejected:
                    return new ProofRejectedException();
                case ErrorCode.AnoncredsCredentialRevoked:
                    return new CredentialRevokedException();
                case ErrorCode.AnoncredsCredDefAlreadyExistsError:
                    return new CredentialDefinitionAlreadyExistsException();
                case ErrorCode.UnknownCryptoTypeError:
                    return new UnknownCryptoTypeException();
                case ErrorCode.WalletItemNotFoundError:
                    return new WalletItemNotFoundException();
                case ErrorCode.WalletItemAlreadyExistsError:
                    return new WalletItemAlreadyExistsException();
                case ErrorCode.WalletQueryError:
                    return new WalletInvalidQueryException();
                case ErrorCode.WalletStorageError:
                    return new WalletStorageException();
                case ErrorCode.WalletDecodingError:
                    return new WalletDecodingException();
                case ErrorCode.WalletEncryptionError:
                    return new WalletEncryptionException();
                case ErrorCode.WalletInputError:
                    return new WalletInputException();
                case ErrorCode.PaymentExtraFundsError:
                    return new ExtraFundsException();
                case ErrorCode.PaymentIncompatibleMethodsError:
                    return new IncompatiblePaymentMethodsException();
                case ErrorCode.PaymentInsufficientFundsError:
                    return new InsufficientFundsException();
                case ErrorCode.PaymentOperationNotSupportedError:
                    return new PaymentOperationNotSupportedException();
                case ErrorCode.PaymentSourceDoesNotExistError:
                    return new PaymentSourceDoesNotExistException();
                case ErrorCode.PaymentUnknownMethodError:
                    return new UnknownPaymentMethodException();

                default:
                    var message = $"An unmapped error with the code '{sdkErrorCode}' was returned by the SDK.";
                    return new IndyException(message, sdkErrorCode);
            }      
        }

        /// <summary>
        /// Gets the error code for the exception.
        /// </summary>
        public int SdkErrorCode { get; private set; }
    }

}
